Client/Frontend/Browser/BrowserViewController.swift

```diff
@@ -2722,6 +2733,125 @@ extension BrowserViewController: ContextMenuHelperDelegate {
+        findInPageContainer.addSubview(view)
+
+        view.snp_makeConstraints { make in
+            make.edges.equalTo(findInPageContainer)
+            make.height.equalTo(UIConstants.ToolbarHeight)
+        }
+
+        self.customSearchEngineButton.snp_remakeConstraints { make in
+            make.width.equalTo(view.snp_height)
+            make.height.equalTo(view.snp_height)
+            make.left.equalTo(view.snp_left).offset(20)
+            make.top.equalTo(view.snp_top)
+        }
+
+        updateViewConstraints()
+        self.findInPageContainer.layoutIfNeeded()
+    }
+
+
+
+    func addCustomSerchEngineForFocusedElement() {
+        guard let webView = tabManager.selectedTab?.webView else {
+            return
+        }
+        webView.evaluateJavaScript("__firefox__.searchQueryForField()") { (result, _) in
+            guard let searchParams = result as? [String: String] else {
+                //Javascript responded with an incorrectly formatted message. Show an error.
+                let alert = ThirdPartySearchAlerts.failedToAddThirdPartySearch()
+                self.presentViewController(alert, animated: true, completion: nil)
+                return
+            }
+            self.addSearchEngine(searchParams)
+            self.customSearchEngineButton.tintColor = UIColor.grayColor()
+            self.customSearchEngineButton.userInteractionEnabled = false
+        }
+    }
+
+    func addSearchEngine(params: [String: String]) {
+        guard let template = params["url"] where template != "",
+            let iconString = params["icon"],
+            let iconURL = NSURL(string: iconString),
+            let url = NSURL(string: template.stringByAddingPercentEncodingWithAllowedCharacters(NSCharacterSet.URLFragmentAllowedCharacterSet())!) else {
+                let alert = ThirdPartySearchAlerts.failedToAddThirdPartySearch()
+                self.presentViewController(alert, animated: true, completion: nil)
+                return
+        }
+
+        let alert = ThirdPartySearchAlerts.addThirdPartySearchEngine { (alert: UIAlertAction) in
+            let shortName = url.domainURL()
+            SDWebImageManager.sharedManager().downloadImageWithURL(iconURL, options: SDWebImageOptions.ContinueInBackground, progress: nil, completed: { [weak self] (img, err, cacheType, success, url) in
+                guard img != nil && shortName.host != nil else {
+                    let alert = ThirdPartySearchAlerts.failedToAddThirdPartySearch()
+                    self?.presentViewController(alert, animated: true, completion: nil)
+                    return
+                }
+                self?.profile.searchEngines.addSearchEngine(OpenSearchEngine(engineId: nil, shortName: shortName.host!, image: img, searchTemplate: template, suggestTemplate: nil, isCustomEngine: true))
+                let Toast = SimpleToast()
+                Toast.showAlertWithText(Strings.ThirdPartySearchEngineAdded)
+
```

Comment:
We should disable the search button here since the "is added" check only works the next time the field is selected. That is, I can currently add the same engine multiple times as long as the field stays focused.